stock_storage_type/models/stock_location.py

```diff
@@ -475,6 +497,55 @@ def select_allowed_locations(
         )
         return valid_locations
 
+    def _order_allowed_locations(self, valid_locations):
+        """Return the ordered list of valid_locations
+
+        By default the order should be the same as self. However, if the
+        valid_locations list contains locations configured to not mix products,
+        we must give priority to locations that already contains products
+        (the ones with less qty first)
+        """
+        valid_no_mix = valid_locations.filtered("do_not_mix_products")
+        qty_by_loc = {}
+        if valid_no_mix:
+            StockQuant = self.env["stock.quant"]
+            domain_quant = [("location_id", "in", valid_no_mix.ids)]
+            qty_by_loc = {
+                item["location_id"][0]: item["quantity"]
+                for item in StockQuant.read_group(
+                    domain_quant,
+                    ["location_id", "quantity"],
+                    ["location_id"],
+                    orderby="id",
+                )
+            }
+        return valid_locations.sorted(self._get_allowed_location_sorter(qty_by_loc))
```

Comment:
```suggestion
         valid_location_ids = set([k for k, qty in qty_by_loc.items() if (float_compare(qty, 0) == 1)] + valid_locations.ids)
         return = self.browse(
             id_ for id_ in self.ids if id_ in valid_location_ids
         )
```

Reply:
@jbaudoux A 'set' is not ordered. Moreover, you lost the new order into the call to self.browse since you iterate on the original order `self.ids'. An other point, the read_group could return 0 if we've negative quants. Finally you must also take into account that the qty into a location could be the same as another one. In such case, the initial order must be used to sort the locations with the same qty. All these cases are covered by the `sorter` method and `_get_allowed_location_sorter` provides an hook if an other addon needs to provide its own logic.

Reply:
@lmignon I updated the code. It wasn't right indeed.
I tried your PR on real customer DB and here is a result of the 2 methods:
len(self.ids): 8183
Without sorting as suggested: 0.001138 
Your version with sorting: 0.402326

Reply:
With real data to stress the logic it's always better... :smirk: